scripts/uosc.lua

```diff
@@ -550,108 +550,388 @@ function get_point_to_rectangle_proximity(point, rect)
+		{0x2CEB0, 0x2EBEF}, -- CJK Unified Ideographs Extension F	 2 SIP	Han
+		{0x2F800, 0x2FA1F}, -- CJK Compatibility Ideographs Supplement	 2 SIP	Han
+		{0x30000, 0x3134F}, -- CJK Unified Ideographs Extension G	 3 TIP	Han
+		{0x31350, 0x323AF}, -- CJK Unified Ideographs Extension H	 3 TIP	Han
+	}
 
----@param text string|number
-function text_length(text)
-	if not text or text == '' then return 0 end
-	local text_length = 0
-	for _, _, length in utf8_iter(tostring(text)) do text_length = text_length + length end
-	return text_length
-end
+	---Get byte count of utf-8 character at index i in str
+	---@param str string
+	---@param i integer?
+	---@return integer
+	local function utf8_char_bytes(str, i)
+		local char_byte = str:byte(i)
+		if char_byte < 0xC0 then return 1
+		elseif char_byte < 0xE0 then return 2
+		elseif char_byte < 0xF0 then return 3
+		elseif char_byte < 0xF8 then return 4
+		else return 1 end
+	end
+
+	---Creates an iterator for an utf-8 encoded string
+	---Iterates over utf-8 characters instead of bytes
+	---@param str string
+	---@return fun(): string
+	local function utf8_iter(str)
+		local byte_start = 1
+		return function()
+			local start = byte_start
+			if #str < start then return nil end
+			local byte_count = utf8_char_bytes(str, start)
+			byte_start = start + byte_count
+			return start, str:sub(start, start + byte_count - 1)
+		end
+	end
 
-function utf8_iter(string)
-	local byte_start, byte_count = 1, 1
+	---Extract Unicode code point from utf-8 character at index i in str
+	---@param str string
+	---@param i integer
+	---@return integer
+	local function utf8_to_unicode(str, i)
+		local byte_count = utf8_char_bytes(str, i)
+		local char_byte = str:byte(i)
+		local unicode = char_byte
+		if byte_count ~= 1 then
+			local shift = 2 ^ (8 - byte_count)
+			char_byte = char_byte - math.floor(0xFF / shift) * shift
+			unicode = char_byte * (2 ^ 6) ^ (byte_count - 1)
+		end
+		for j = 2, byte_count do
+			char_byte = str:byte(i + j - 1) - 0x80
+			unicode = unicode + char_byte * (2 ^ 6) ^ (byte_count - j)
+		end
+		return round(unicode)
+	end
 
-	return function()
-		if #string < byte_start then return nil end
+	---Convert Unicode code point to utf-8 string
+	---@param unicode integer
+	---@return string?
+	local function unicode_to_utf8(unicode)
+		if unicode < 0x80 then return string.char(unicode)
+		else
+			local byte_count
+			if unicode < 0x800 then byte_count = 2
+			elseif unicode < 0x10000 then byte_count = 3
+			elseif unicode < 0x110000 then byte_count = 4
+			else return end -- too big
+
+			local res = {}
+			local shift = 2 ^ 6
+			local after_shift = unicode
+			for _ = byte_count, 2, -1 do
+				local before_shift = after_shift
+				after_shift = math.floor(before_shift / shift)
+				table.insert(res, 1, before_shift - after_shift * shift + 0x80)
+			end
+			shift = 2 ^ (8 - byte_count)
+			table.insert(res, 1, after_shift + math.floor(0xFF / shift) * shift)
+			return string.char(unpack(res))
```

Comment:
That already happens in [defaults.lua](https://github.com/mpv-player/mpv/blob/master/player/lua/defaults.lua#L2)

Reply:
If the `defaults.lua` was a thing for long enough so that we don't have to be afraid that someone's mpv doesn't have it, than sure keep it at `unpack()` only, but at least add the `@diagnostic` line so that the editor doesn't yell at me.

Reply:
Blame shows it's been there for 8 years. Will that be enough? :^)

Reply:
Yeah it is :) I was just trying to be consistent with #155. I guess it wasn't needed after all. But we still need the diagnostic thingy or this turns into error/warning.